### DISCUSSION TITLE
How to have a common authentication-service for centralise the auth flow?

### DISCUSSION THREAD

**Discussion by ghost** [mia-platform/community#303] [Q&A] [answered]
### Which section?

Console

### Which is the version?

v11.2.0

### Description

Hi all!

I would like to have a common **authentication-service** to handle the auth process and the integration with the IDP, with every **authorization-service** in every projects to point to this common one.

I'm struggling right now because I'm getting a "redirect loop" from the `/web-login` path and it seems that the `authorization-service` is not able to contact the common `authentication-service`. 

My configuration is as follows:
- the common project has _api-gateway_, _authorization-service_, _authentication-service_ and _login-site_ services already configured and working on their own;
- the non-common project has an _api-gateway_, _authorization-service_ and _login-site_ services. The _authorization-service_ is configured to call the common _authentication-service_ without passing through the common _api-gateway_. I've setup a basic rule in the Advanced section of the console to execute a redirect when I get a text/html body, but this seems to not working properly.

So a few questions:
- can I achieve this?
- how do I need to configure the "non-common" projects in order to make everything working properly?

Thank you very much!

**Comment by davidebianchi** [comment]
Hi Ivan! 

I have a couple of question to better understand the context.

Why there are multiple login-site? Do you want a single login or all the different projects should authenticate with a custom login page?

To verify that the user have a correct session, you can use directly envoy with the jwt filter if you use the authentication-service (here [the docs](https://docs.mia-platform.eu/docs/development_suite/api-console/advanced-section/api-gateway-envoy/jwt-authn)) and the authorization-service could consider the header passed in it as trusted.
In this case, the authentication service should only manage the authentication flows. Could this path help you?

**Reply by ghost** [reply]
Hi Davide, thanks for the response!

I've got multiple login-site just for a config duplication, so we started from a working point and try to replicate the configuration but on multiple projects. We are looking for a single login-site, so I think at this point that we only need the login-site on the common project, right?

We are going to try the envoy configuration proposed, but with nginx is there anything that we could do? How do we need to configure the files in the Advanced section to handle correctly the redirect to / from the login-site?

Thank you!

**Reply by davidebianchi** [reply]
With nginx, you can:

in the `backoffice-root-location-extension.conf` extension file add:

```nginx
error_page 401 = /error401.$extension_by_http_accept;
```

and in `backoffice-server-extension.conf` add (you should change the PATH_TO_LOGIN_SITE accordingly):

```nginx
location /error401.html {
  include /etc/nginx/customization.d/header-debug.conf;

  return 302 '$original_request_scheme://$original_request_host/PATH_TO_LOGIN_SITE;
}

location /error401.json {
  include /etc/nginx/customization.d/header-debug.conf;

  return 401 $content_401;
}
```

There is also this documentation page which could help you to configure api-gateway authorization in a multi gateway architecture [docs](https://docs.mia-platform.eu/docs/development_suite/api-console/advanced-section/api-gateway/how-to#how-to-manage-authorization-in-a-multi-gateway-architecture).

**Reply by ghost** [reply, marked as answer]
Hi Davide! 

Thank you for the advice. I've came across another problem: the configuration of the _authentication-service_ requires the `redirectUrl` field, that must match the one configured inside the IDP of choosing. The problem is that this parameters can't be "dynamic" and for my purposes, I would need that the host of this URL points to the called host by the client.

So to solve this issue, I think that the best solution is having the centralised _authentication-service_ at the same level of the Gateway, resolve the auth flow once, and then redirect the call to the correct project via Console's Proxies with the appropriate Headers set, just like mentioned in the link of the docs you have posted.

Could you confirm me this idea?
Thank you very much

**Reply by davidebianchi** [reply]
Yes, probably this is the best way to follow
